pylib/mozhginfo/mozhginfo/pushlog_client.py

```diff
@@ -0,0 +1,187 @@
+        # not guaranteed (due to system clock skew)
+        # We can interact with self-serve with the 12 char representation
+        revisions.append(pushes[str(push_id)]["changesets"][0])
+
+    return revisions
+
+
+def query_revisions_range_from_revision_before_and_after(repo_url, revision, before, after):
+    """
+    Get the start and end revisions based on the number of revisions before and after.
+
+    Raises PushlogError if pushlog data cannot be retrieved.
+    """
+    try:
+        push_info = query_revision_info(repo_url, revision)
+        pushid = int(push_info["pushid"])
+        start_id = pushid - before
+        end_id = pushid + after
+        revlist = query_pushid_range(repo_url, start_id, end_id)
+    except:
+        raise PushlogError('Unable to retrieve pushlog data. '
+                           'Please check repo_url %s and revision %s specified.'
+                           % (repo_url, revision))
+
+    return revlist
+
+
+def query_revision_info(repo_url, revision, full=False):
+    """
+    Return a dictionary with meta-data about a push including:
+
+        * changesets
+        * date
+        * user
+    """
+    url = "%s?changeset=%s&tipsonly=1" % (JSON_PUSHES % {"repo_url": repo_url}, revision)
+    if full:
+        url += "&full=1"
+    LOG.debug("About to fetch %s" % url)
+    req = requests.get(url)
+    data = req.json()
+    assert len(data) == 1, "We should only have information about one push"
+    push_id, push_info = data.popitem()
+    push_info["pushid"] = push_id
+    if not full:
+        LOG.debug("Push info: %s" % str(push_info))
+    else:
+        LOG.debug("Requesting the info with full=1 can yield too much unnecessary output "
+                  "to debug anything properly")
+    return push_info
+
+
+def query_full_revision_info(repo_url, revision):
+    """
+    Return the full 40 chars revision number which correspond to the 12 chars
+    revision number we given, and throw an error out if found collision in there.
+
+    :param repo_url: The repo url of a repository e.g. mozilla-inbound
+    :param revision: push revision (40 chars)
+    :type revision: str
+    :return:
+    """
+    revision = query_revision_info(repo_url, revision)['changesets'][0]
+    if not revision:
+        raise PushlogError('Unable to retrieve pushlog info by repo_url: %s'
+                           ' and revision: %s. Please check repo_url and revision secified.'
+                           % (repo_url, revision))
+    return revision
+
+
+def query_repo_tip(repo_url):
+    """Return the tip of a branch."""
+    url = "%s?tipsonly=1" % (JSON_PUSHES % {"repo_url": repo_url})
+    recent_commits = requests.get(url).json()
+    tip_id = sorted(map(int, recent_commits.keys()))[-1]
+    return recent_commits[str(tip_id)]["changesets"][0]
+
+
+def valid_revision(repo_url, revision):
+    """Verify that a revision exists in a given branch."""
```

Comment:
I think you mean "in the given repository." Mozilla uses "branch" to refer to separate Mercurial repositories. It's wrong and confusing. But it's difficult to change popular nomenclature.